backends/cuda/ceed-cuda-operator.c

```diff
@@ -403,34 +491,185 @@ static int CeedOperatorApplyAdd_Cuda(CeedOperator op, CeedVector invec,
+                                          numactivein*numactiveout*numelements*Q,
+                                          strides, rstr); CeedChk(ierr);
+  // Create assembled vector
+  ierr = CeedVectorCreate(ceedparent, numelements*Q*numactivein*numactiveout,
+                          assembled); CeedChk(ierr);
+  ierr = CeedVectorSetValue(*assembled, 0.0); CeedChk(ierr);
+  CeedVector assembledE;
+  if (impl->eandqdiffer) {
+    ierr = CeedVectorCreate(ceedparent, numelements*Q*numactivein*numactiveout,
+                            &assembledE); CeedChk(ierr);
+    ierr = CeedVectorGetArray(assembledE, CEED_MEM_DEVICE, &a); CeedChk(ierr);
+  } else {
+    ierr = CeedVectorGetArray(*assembled, CEED_MEM_DEVICE, &a); CeedChk(ierr);
+  }
+
+  // Input basis apply
+  ierr = CeedOperatorInputBasis_Cuda(numelements, qfinputfields, opinputfields,
+                                     numinputfields, true, impl);
+  CeedChk(ierr);
+
+  // Assemble QFunction
+  for (CeedInt in=0; in<numactivein; in++) {
+    // Set Inputs
+    ierr = CeedVectorSetValue(activein[in], 1.0); CeedChk(ierr);
+    if (numactivein > 1) {
+      ierr = CeedVectorSetValue(activein[(in+numactivein-1)%numactivein],
+                                0.0); CeedChk(ierr);
+    }
+    // Set Outputs
+    for (CeedInt out=0; out<numoutputfields; out++) {
+      // Get output vector
+      ierr = CeedOperatorFieldGetVector(opoutputfields[out], &vec);
+      CeedChk(ierr);
+      // Check if active output
+      if (vec == CEED_VECTOR_ACTIVE) {
+        CeedVectorSetArray(impl->qvecsout[out], CEED_MEM_DEVICE,
+                           CEED_USE_POINTER, a); CeedChk(ierr);
+        ierr = CeedQFunctionFieldGetSize(qfoutputfields[out], &size);
+        CeedChk(ierr);
+        a += size*Q*numelements; // Advance the pointer by the size of the output
+      }
+    }
+    // Apply QFunction
+    ierr = CeedQFunctionApply(qf, Q*numelements, impl->qvecsin, impl->qvecsout);
+    CeedChk(ierr);
+  }
+
+
+  // Un-set output Qvecs to prevent accidental overwrite of Assembled
+  for (CeedInt out=0; out<numoutputfields; out++) {
+    // Get output vector
+    ierr = CeedOperatorFieldGetVector(opoutputfields[out], &vec);
+    CeedChk(ierr);
+    // Check if active output
+    if (vec == CEED_VECTOR_ACTIVE) {
+      CeedVectorSetArray(impl->qvecsout[out], CEED_MEM_HOST, CEED_COPY_VALUES,
+                         NULL); CeedChk(ierr);
+    }
+  }
+
+  // Restore input arrays
+  ierr = CeedOperatorRestoreInputs_Cuda(numinputfields, qfinputfields,
+                                        opinputfields, true, impl);
+  CeedChk(ierr);
+
+  // Restore output
+  if (impl->eandqdiffer) {
+    ierr = CeedVectorRestoreArray(assembledE, &a); CeedChk(ierr);
+    ierr = CeedElemRestrictionApply(*rstr, CEED_TRANSPOSE, assembledE,
+                                    *assembled, request); CeedChk(ierr);
+    ierr = CeedVectorDestroy(&assembledE); CeedChk(ierr);
```

Comment:
Now I see what you're saying. That would explain why these lines are here when they are not in the CPU impl. CPU does not have this issue because our Evector and Qvector layouts are the same.

@YohannDudouit thoughts? I feel like rearanging the cuda ref/reg/shared backends to have the same Lvector and Qvector layout would take a fair bit of effort but would make things a lot cleaner and provide correct output when using user provided qdata with a non backend strides restriction.